backend/src/tasks/shodan.ts

```diff
@@ -0,0 +1,126 @@
+  country_name: string;
+  domains: string[];
+  hostnames: string[];
+  org: string;
+  ports: number[];
+  os: string;
+  city: string;
+  longitude: number;
+  latitude: number;
+  data: {
+    port: number;
+    product: string;
+    data: string;
+    cpe: string[];
+    version: string;
+    vulns: {
+      [title: string]: {
+        verified: boolean;
+        references: string[];
+        cvss: number;
+        summary: string;
+      };
+    };
+  }[];
+}
+
+const sleep = (milliseconds) => {
+  return new Promise((resolve) => setTimeout(resolve, milliseconds));
+};
+
+export const handler = async (commandOptions: CommandOptions) => {
+  const { organizationId, organizationName } = commandOptions;
+
+  console.log('Running shodan on organization', organizationName);
+
+  const domainsWithIPs = await getIps(organizationId);
+
+  const ipToDomainsMap: IpToDomainsMap = domainsWithIPs.reduce<IpToDomainsMap>(
+    (map: IpToDomainsMap, domain: Domain) => {
+      if (!map[domain.ip]) {
+        map[domain.ip] = [];
+      }
+      map[domain.ip].push(domain);
+      return map;
+    },
+    {}
+  );
+
+  const chunks = chunk(domainsWithIPs, CHUNK_SIZE);
+
+  for (const domainChunk of chunks) {
+    let { data } = await axios.get<ShodanResponse[]>(
+      `https://api.shodan.io/shodan/host/${domainChunk
+        .map((domain) => domain.ip)
+        .join(',')}?key=${process.env.SHODAN_API_KEY}`
+    );
+
+    // If only one item is returned, the response will not be an array
+    if (!Array.isArray(data)) {
+      data = [data];
+    }
+    for (const item of data) {
+      const domains = ipToDomainsMap[item.ip_str];
+      for (const domain of domains) {
+        for (const service of item.data) {
+          const [serviceId] = await saveServicesToDb([
+            plainToClass(Service, {
+              domain: domain,
+              discoveredBy: { id: commandOptions.scanId },
+              port: service.port,
+              lastSeen: new Date(Date.now()),
+              banner: sanitizeStringField(service.data),
+              shodanResults: {
+                product: service.product,
+                version: service.version,
+                cpe: service.cpe
+              }
+            })
+          ]);
+          if (service.vulns) {
+            const vulns: Vulnerability[] = [];
+            for (const cve in service.vulns) {
+              vulns.push(
+                plainToClass(Vulnerability, {
+                  domain: domain,
+                  lastSeen: new Date(Date.now()),
+                  title: cve,
+                  cve: cve,
+                  cpe:
+                    service.cpe && service.cpe.length > 0
+                      ? service.cpe[0]
+                      : null,
+                  cvss: service.vulns[cve].cvss,
+                  state: 'open',
+                  source: 'shodan',
+                  description: service.vulns[cve].summary,
+                  needsPopulation: true,
```

Comment:
Should this be false? Otherwise we might end up overwriting the description from shodan with our own description (or maybe we want to do this?)

Reply:
This is intentionally true, so that we populate it with our own description and references. I think it's useful to have this all from the same source.